Show current progress

james-garner-canonical opened the pull request:
This pull request shows current progress on this charm for `calibre-web`.

See `README.md` for details on how to run, actions, etc.

I wanted to integrate with cos-lite and query the `calibre-web` logs with `loki` via `grafana` but bailed on that plan after sinking too much time into trying to get anything to show up in `grafana`, so currently you can just deploy `calibre-web`, write and query an ebook library, and configure how writing works (see `README.md`).

Integration tests currently cover pretty much everything, I think.

Unit tests are in a much more minimal state, and will need to be updated. There are a couple of tests using harness. I'd like to try using scenario for some unit tests too, but I haven't gotten to that yet.

james-garner-canonical committed:
Add my working code that deploys calibre-web

james-garner-canonical committed:
Add unmodified project files provided by charmcraft init

james-garner-canonical committed:
Add storage volume

james-garner-canonical committed:
Add storage and the ability to pass your Calibre Library as a resource

Also package the default starter calibre library to use as the default
for calibre-web if the user doesn't supply anything

james-garner-canonical committed:
Slight refactor and better error handling

james-garner-canonical committed:
Support configurable options for handling existing library

james-garner-canonical committed:
Add some actions

james-garner-canonical committed:
Minor cleanup

james-garner-canonical committed:
Test for commands before installing dependencies and rename action

For consistency

james-garner-canonical committed:
Quick and dirty check in status collected event for blocked

james-garner-canonical committed:
Move some definitions to better places

james-garner-canonical committed:
Handle bad user input better

james-garner-canonical committed:
Refactor file moving code into helper method

james-garner-canonical committed:
Rename to calibre-web instead of calibre-web-charm

james-garner-canonical committed:
Rename / add parameters

james-garner-canonical committed:
Better typing and variable use for pebble dict

james-garner-canonical committed:
Remove commente out code

james-garner-canonical committed:
Rearrange method order to put observers first

james-garner-canonical committed:
Remove unrelated skeleton tests and update pebble ready test

james-garner-canonical committed:
Add build integration test (deploy test not yet working)

james-garner-canonical committed:
Remove unused code

james-garner-canonical committed:
Add note about simplifying logic

james-garner-canonical committed:
Remove typing feature requiring python > 3.10

james-garner-canonical committed:
Don't immediately write library on config update

james-garner-canonical committed:
Rename variables to reflect action name

james-garner-canonical committed:
Remove unused comments

james-garner-canonical committed:
Cleanup docstring, header, move logging setup to after classes

james-garner-canonical committed:
Update docstrings

james-garner-canonical committed:
Add build and deploy smoke test

Passes in VM. Fails sometimes on local due to juju storage attaching
shenanigans.

james-garner-canonical committed:
Small changes to deploy test

james-garner-canonical committed:
Add status test for updating config

james-garner-canonical committed:
Add integration tests for library-write action

Also stop using raise_on_blocked as it produces spurious errors when the
previous status was intentionally set to blocked.

james-garner-canonical committed:
Use a variable for sentinel path

james-garner-canonical committed:
Update project config (coverage, ruff, etc)

james-garner-canonical committed:
Rename some variables for clarity

james-garner-canonical committed:
Run update_config in unit tests

Since this doesn't trigger a library-write anymore

james-garner-canonical committed:
Add more comprehensive integration testing

james-garner-canonical committed:
Run ruff format

james-garner-canonical committed:
Remove examples.txt

james-garner-canonical committed:
Move pebble_ready ActiveStatus to _on_collect_status

Invoke harness.evaluate_status in unit tests

james-garner-canonical committed:
Update readme

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+    def _on_library_info(self, event: ops.ActionEvent) -> None:
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        format = cast(str, event.params[LIBRARY_INFO_FORMAT_PARAM])
+        match format:
```
This won't work in 20.04, but I assume you don't care about that?

james-garner-canonical replied:
I should explicitly set the expected python version ... It was nice to use some more modern python features and not worry about python 3.8 compatibility for a while. I suppose it would be worth supporting 20.04 if I was to continue with this charm though

tonyandrewmeyer commented on tox.ini:
```diff
@@ -0,0 +1,86 @@
+# Copyright 2024 tmp
```
Not sure if you're planning to continue with this charm after the onboarding process, but if so I assume you would want a different author name here :smile:.

tonyandrewmeyer commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+
+
+# (Recommended)
+title: Charm Template
```
I'd put something meaningful here or remove this.

tonyandrewmeyer commented on spread.yaml:
```diff
@@ -0,0 +1,152 @@
+project: calibre-web-charm-k8s-tests
```
I'm assuming this is unchanged from the profile. If it is changed and you'd like it reviewed, let me know :)

james-garner-canonical replied:
Sorry, I did commit a bunch of stuff provided by `charmcraft init` that ultimately went unused. I should have removed it before asking for review. Glad you caught that it wasn't changed.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,47 @@
+    assert harness.model.unit.status == ops.ActiveStatus()
+
+
+def test_config_changed(harness: ops.testing.Harness[CalibreWebCharm]):
```
It feels like this would be slightly better as a parametrized test, but that's only a weak opinion.

james-garner-canonical replied:
I think you're right.

dimaqq commented on README.md:
```diff
@@ -0,0 +1,60 @@
+* detailed contribution guide > documentation or CONTRIBUTING.md
+
+Use links instead.
+-->
```
you'd want to remove the comments, at least when the readme is ready.

dimaqq commented on README.md:
````diff
@@ -0,0 +1,60 @@
+```bash
+juju deploy ./calibre-web_ubuntu-22.04-amd64.charm \
+    --resource calibre-web-image=lscr.io/linuxserver/calibre-web:latest \
+    --resource calibre-library=./empty.zip  # an empty file
````
I'd rather make the resource optional. I think that's possible, isn't it?

james-garner-canonical replied:
As far as I can tell it isn't possible to make a resource strictly optional -- let me know if I'm wrong here. But what you can do is to provide a default resource via charmhub. I got the idea to use an empty file as the charmhub provided default from [this charm](https://github.com/canonical/openstack-exporter-operator/blob/main/charmcraft.yaml).

dimaqq commented on README.md:
````diff
@@ -0,0 +1,60 @@
+
+The user's library can be provided as a resource, at deployment, or e.g.
+```bash
+juju attach-resource calibre-web calibre-library=$ZIPPED_USER_LIBRARY
````
Can the library be updated by the workload?

james-garner-canonical replied:
From the calibre-web gui you can add new books to the library or point it to a totally different library. Is that what you're asking?

I think ideally the user would be able to provide their own (persistent) library volume, which could eliminate the need for the calibre-library resource.

dimaqq commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+
+
+# (Required)
+summary: A very short one-line summary of the charm.
```
you may want to change these 😆

dimaqq commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+    location: /books
+    minimum-size: 1G
+    properties:
+      - transient
```
I'm curious why.

james-garner-canonical replied:
If you're asking about the use of `transient`, the real reason is that the juju "how to" on [adding storage](https://juju.is/docs/sdk/use-storage-in-a-charm) uses this setting, and the [charmcraft.yaml documentation](https://juju.is/docs/sdk/charmcraft-yaml#heading--storage) doesn't really explain it, so I just left it as `transient`. I'd be grateful if you could point me at a good explanation, it probably exists and I've just overlooked it while trying to get the charm working.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
+        self.unit.set_ports(8083)
```
I'm pretty sure other charms do this too, but I feel like interaction with Juju should be minimal in `__init__`. I would put this in `install` or `start`.

james-garner-canonical replied:
Today I learned about the `install` and `start` events, thanks! I'll look into using them in this charm, I think they're a good fit for a few things. Are they in the current tutorial? Might be a good addition if not.

tonyandrewmeyer replied:
They're in the machine tutorial, because install is most common with machine charms where you are potentially installing a bunch of packages rather than having a prepared image with everything ready.

For k8s, there's the timing/sequencing issue with pebble-ready, because most of the stuff you'd do in start is going to also need the container (ports being an exception). So you probably want a holistic handler than bundles config-changed, relation-changed, pebble-ready, secret-changed, etc rather than start. I think we likely need to  a better job of guiding towards this in the docs.

dimaqq commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+#    organize:
+#      "libxxx.so*": lib/
+#    prime:
+#      - lib/
```
pls clean up.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
```
Is there any point observing this if you don't do anything?

james-garner-canonical replied:
Does `collect-status` get triggered on every event, regardless of whether anything is observed? If so, then this can be removed. It's really just there because it used to do something, and I kept it around in case it started doing something again.

dimaqq commented on pyproject.toml:
```diff
@@ -0,0 +1,44 @@
+# Testing tools configuration
```
It would be nice to set target Python version here.

Although it can be inferred from supported bases, most tooling, like ruff or pyright wouldn't know that.

dimaqq commented on spread.yaml:
```diff
@@ -0,0 +1,152 @@
+project: calibre-web-charm-k8s-tests
```
TIL about spread...

I'm unclear how it's invoked though. Maybe some files are missing from the PR?

james-garner-canonical replied:
Sorry, I committed a bunch of files that were provided by `charmcraft init`, and as far as I can tell `test/spread` isn't used at all at the moment, so I'll remove those files.

tonyandrewmeyer replied:
FYI, the spread files power `charmcraft test`. It's currently very poorly documented and not really clear where this is meant to head (it's on the list for discussion in The Hague).

james-garner-canonical replied:
Sorry for the repetition, but -- I had no idea about `charmcraft test`!

I think I'll remove the spread stuff for now since I've just been using `tox` anyway.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+    def _get_library_write_behaviour(
+        self,
+    ) -> tuple[LibraryWriteBehaviour, ops.ActiveStatus] | tuple[None, ops.BlockedStatus]:
+        library_write_behaviour = self.model.config[LIBRARY_WRITE_CONFIG]
```
I would normally use `self.config` for the brevity.

james-garner-canonical replied:
TIL, thanks!

dimaqq commented on spread.yaml:
```diff
@@ -0,0 +1,152 @@
+    type: adhoc
+    allocate: |
+      # Mitigate issues found when launching multiple mutipass instances
+      # concurrently. See https://github.com/canonical/multipass/issues/3336
```
that issues has been closed as could not be reproduced...

dimaqq commented on spread.yaml:
```diff
@@ -0,0 +1,152 @@
+      # concurrently. See https://github.com/canonical/multipass/issues/3336
+      sleep 0.$RANDOM
+      sleep 0.$RANDOM
+      sleep 0.$RANDOM
```
Why thrice? Would `sleep 1.$RANDOM` not do essentially the same?

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+
+    def _get_library_write_behaviour(
+        self,
+    ) -> tuple[LibraryWriteBehaviour, ops.ActiveStatus] | tuple[None, ops.BlockedStatus]:
```
This is probably just taste, but I would use a simpler return value here of just the `LibraryWriteBehaviour` and raise `ValueError` if the config is invalid.

In general, there's a plan to make it simpler for charms to validate and apply typing to config (and action params and maybe relation data) next cycle, that would clean this up a bit.

james-garner-canonical replied:
It is a bit ugly looking! I borrowed from golang's idiom for error handling here in my attempt to keep the status-evaluating and value-returning code in one place -- it's called both in the code that needs the config value (or to abort if it's invalid), and in the `collect-status` observer to add the status.

I originally threw an exception, but that made it tricky to get the bad config value for the `blocked` message. I mean you could attach it to the `Exception`, but that feels a bit nasty.

An alternative with a nicer return value annotation would be `LibraryWriteBehaviour | ops.BlockedStatus`, but it would make the code at the two calling sites (imo) a bit uglier ...

The planned changes you mention would definitely help with this case, but it seems like keeping your status evaluations dry / single source of truth is still a bit tricky, and I'd like to learn or figure out the best idioms there.

james-garner-canonical replied:
I use this in three places currently:
```python
def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
    ...
    _, status = self._get_library_write_behaviour()
    event.add_status(status)

def _on_library_write(self, event: ops.ActionEvent) -> None:
    library_write_behaviour, status = self._get_library_write_behaviour()
    if library_write_behaviour is None:
        event.fail(status.message)
        return
    self._push_library_to_storage()  # also triggered during initial setup
    event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})

def _push_library_to_storage(self) -> None:
    library_write_behaviour, _ = self._get_library_write_behaviour()
    if library_write_behaviour is None:
        return
    ...
    match library_write_behaviour:
        ...
```
I guess I'm wondering if this is a good pattern for handling this kind of thing? Is there a more preferred idiom currently? Or is the fact that I'm reaching for a solution like this a sign that I've taken a wrong turn somewhere?

tonyandrewmeyer replied:
> It is a bit ugly looking! I borrowed from golang's idiom for error handling here

Indeed, it's ugly exactly like Go :stuck_out_tongue: .

> I originally threw an exception, but that made it tricky to get the bad config value for the `blocked` message. I mean you could attach it to the `Exception`, but that feels a bit nasty.

This doesn't seem nasty?

```python
def get_config(self):
    raise ValueError(f"your {value} is bad, dummy")

def use_config(self):
    try:
        config = get_config()
    except ValueError as e:
        event.fail(str(e))  # or event.fail(e.args[0])
```

> An alternative with a nicer return value annotation would be `LibraryWriteBehaviour | ops.BlockedStatus`, but it would make the code at the two calling sites (imo) a bit uglier ...

I'm definitely -1 on having a return type that's a mix of two entirely different types (except for `None` and something).

> The planned changes you mention would definitely help with this case, but it seems like keeping your status evaluations dry / single source of truth is still a bit tricky, and I'd like to learn or figure out the best idioms there.

I would do something like this:

```python
def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
    ...
    try:
        self._get_library_write_behaviour()  # Ignore return value, we just want to validate.
    except ValueError as e:
        event.add_status(ops.BlockedStatus(str(e))

def _on_library_write(self, event: ops.ActionEvent) -> None:
    try:
        library_write_behaviour self._get_library_write_behaviour()
    except ValueError as e:
        event.fail(str(e))
        return
    self._push_library_to_storage()  # also triggered during initial setup
    event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})

def _push_library_to_storage(self) -> None:
    try:
        library_write_behaviour = self._get_library_write_behaviour()
    except ValueError:
        return
    ...
    match library_write_behaviour:
        ...
```

james-garner-canonical replied:
Your proposed solution is probably idea in this case.

But while we're here, how would you handle things if a method could result in (say) either a `blocked` or `waiting` status? Would the ugly golang idiom be better here?
```python
def f(self) -> tuple[Value, None] | tuple[None, BadStatus]:
```
Or maybe a mapping of (custom?) exception types to statuses, either generally or just in a series of `except` blocks in `_on_collect_status`?

tonyandrewmeyer replied:
I've seen people do something like:

```
class InvalidConfigError(Exception);
    def __init__(self, message, status):
        self.message = message
        self.status = status

class MyCharm(...):
    ...
    def validate(self):
        if something:
            raise InvalidConfig("message for logging", ops.BlockedStatus("blah"))
```

There are other people (maybe all of them!) in our team that like Go more than me, so might be more inclined to the (value, err) style, although it still feels unpythonic to me.

I think in most cases you actually only expect to have a specific status - if you're getting config, then I'd expect that you end up in blocked, if you are getting relation data I'd expect you end up in Waiting, etc. So I think you probably do end up just raising different exception types (probably still custom ones) depending on the situation.

james-garner-canonical replied:
Thanks for the explanation, Tony, that's really helpful. I think of myself as more of a python guy, but transferring information around on exceptions does feel a little bit off to me compared to returning the information directly ...

Here's a wild idea for your entertainment - make the status classes all inherit from `Exception` (via some lineage) so that you can do:
```python
try:
    something()
except (BlockedStatus, MaintenanceStatus, WaitingStatus) as e:
    event.add_status(e)
```

How to interpret `raise ops.ActiveStatus()` is left as an exercise for the reader.

tonyandrewmeyer replied:
Ha, I could see all but active as exceptions, where somewhere in the framework they are caught and set the status and provide a way to jump out of the code. In another timeline ...

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        self.lines.append(stuff)
+
+
+logger = logging.getLogger(__name__)
```
This would customarily be near the top of the module, although it doesn't really matter.

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
```
Why parametrise these? Is that only to get better callback types?

james-garner-canonical replied:
This just seems to me like the right way to handle string values coming from `charmcraft.yaml`. What do you think?

Benefits I found while working on this charm were ease of renaming values in both `charmcraft.yaml` and `src/charm.py`, and not having to type the literal string values in my tests (and keep them up to date too).

I'm also hopeful that I can make changes to `ops` that will let the `on[NAME].event` form provide exact event types, which doesn't seem like it would be possible with `on.name_event`.

james-garner-canonical replied:
I actually feel like it would be kind of neat if ops provided a nice way to pull the values directly from the `charmcraft.yaml` and really have a single source of truth, but that's probably a lot of over engineering for what is probably a complete non-issue in practice.

tonyandrewmeyer replied:
Or if charmcraft could pull your actions and config from the charm code to populate actions.yaml and config.yaml :)

This ties a bit into the discussion and work we have planned around config/action/relation types, having a useful single source of truth for those, etc. Before we get to that we can hopefully at least lint that there aren't mismatches.

In terms of using `on[name]` vs `on.name_`, I'm mostly convinced, assuming that the improvements do come along.

In terms of `on[NAME_SET_TO_FOO]` vs `on['foo']`, I guess it does reduce repetition (particularly in tests?) and so make it easier to rename, but the plain string looks cleaner to me - maybe I'm just used to it?

james-garner-canonical replied:
> Or if charmcraft could pull your actions and config from the charm code to populate actions.yaml and config.yaml :)

That would be cool!

> In terms of on[NAME_SET_TO_FOO] vs on['foo'], I guess it does reduce repetition (particularly in tests?) and so make it easier to rename, but the plain string looks cleaner to me - maybe I'm just used to it?

Definitely the fact that I went with `on[LIBRARY_INFO_ACTION].action` rather than `on[LIBRARY_INFO].action` makes this uglier than perhaps it needs to be.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})
+
+    def _on_library_info(self, event: ops.ActionEvent) -> None:
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
```
```suggestion
        container = self.unit.get_container(CONTAINER_NAME)
```

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                try:
+                    container.exec(['which', 'tree']).wait()
+                except ops.pebble.ExecError:
+                    logger.debug('_on_library_info: tree: installing dependencies')
```
It feels like installing dependencies would be better in a `pebble-ready` hook, although I guess if no-one ever uses this option then it avoids needing to install them.

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                SERVICE_NAME: {
+                    'override': 'replace',
+                    'summary': 'calibre-web',
+                    'command': command,
```
Command can be a list too.

james-garner-canonical replied:
That would be very nice, if so! `ops.pebble.ServiceDict`  has `'command'` typed as a `str` in the `TypedDict` definition, but maybe you can just ignore the linter errors and pass a list in anyway and it's fine at runtime?

Note: it seems it can't be a list (as Dima noted in chat).

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        """Return a dictionary representing the Pebble layer."""
+        c = ' && '.join(
+            [
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
+                'cd /app/calibre-web',
+                'python3 /app/calibre-web/cps.py',
```
service env can be set in the pebble layer
service `working-dir` can also be set there

I'm not sure what `bash .../run` does, but I wonder if it makes sense to pull that out. That really depends on whether you'd expect it to change from one workload / container image version to the next.

james-garner-canonical replied:
The `bash .../run` line runs a script that would normally be run by `s6-overlay` when deploying this docker image. It creates some files and directories and sets some permissions. Now that you mention it, I guess this charm only targets this version of the docker image, since there are a whole bunch of other places the authors could add various important things that `s6-overlay` would be expected to run. (It refuses to run in this case because it's not pid 1 (pebble is)). I think in the long term I'd ideally just create a new Dockerfile for the charm without any of the `s6` stuff.

Tony also pointed out that env and working-dir can be handled in the pebble layer, I'll try that.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+class CaptureStdOut:
+    """Capture stdout when executing processes.
+
+    The default stdout stream is often closed when we try to read it after a process completes.
```
It seems odd that you should need this - if we can't reliably get output from an exec call that would seem to be an ops/Pebble bug. Did it happen when you didn't give it a stdout and used `wait_output`? Or you were always streaming?

james-garner-canonical replied:
Ah I see where I went wrong, I just used `process.wait()`, and then when I tried to read `process.stdout` the stream would be closed. I'd already run into the issue when the process would raise an exception, and made `CaptureStdout` for debugging purposes (capturing both stdout and stderr), so it was apparently easier to just clean it up a little and use it than to discover `wait_output`. I'm assuming that `wait_output` will let me get ride of `CaptureStdout`, thanks!

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'tree': '\n'.join(stdout.lines)})
```
There'a a limit on juju action result size.

james-garner-canonical replied:
I ran into this when trying to provide a zipped library as one of the output options. I'll have to look into when you'd hit this limit with `tree` and how to handle that error if so (perhaps preemptively truncating tree output or running again with a depth limit).

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                    f'Invalid value {format} for {LIBRARY_INFO_FORMAT_PARAM} parameter'
+                    f' of {LIBRARY_INFO_ACTION} action.'
+                )
+                logger.error(f'_on_library_info: {format}: {msg}')
```
You probably don't need `{format}` in here since it's inside of `msg`.

Also, it's your charm so ignore this if you like, but you shouldn't really use f-strings in logging calls. If it's written like:

```suggestion
                logger.error('_on_library_info: %s', msg)
```

Then logging handlers can be smarter in what they do, like aggregate errors based on the raw message, or skip evaluating the args (which wouldn't do anything with this plain string, of course).

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        """Define and start the workload using the Pebble API."""
+        container = event.workload
+        container.add_layer(SERVICE_NAME, {**self.get_pebble_layer()}, combine=True)
```
unnecessary `{**}`, I think.

james-garner-canonical replied:
This is just to make the type checker happy, but it's not obvious at all that that's why it's done, is it?

My `get_pebble_layer` returns a `TypedDict` with the guaranteed keys that I provide. When unpacking like this, the type checker is happy to convert that into a `ops.pebble.LayerDict`, which has all optional keys (a superset of the keys I provide).

This is safer than a `cast`, since it will catch incompatibilities introduced by any changes to the `ops.pebble.LayerDict` definition, or errors introduced in `get_pebble_layer`, which a `cast` or `type: ignore` would mask.

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+    def _on_storage_attached(self, event: ops.StorageAttachedEvent) -> None:
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
+
+        Bad config values are handled in _on_collect_status
+        """
+        pass
```
That's a bit weird.
I guess you are observing this event, just so that collect-status would run, aren't you?

If config did change, shouldn't the workload be updated or restarted?

dimaqq replied:
Or, if the workload doesn't actually care about the config, and the config is only used by the export action, would it be better to:

* either eval config in the action only,
* or make the config field an action parameter instead?

james-garner-canonical replied:
The only reason that the `config` item couldn't just be a parameter for the `library-write` action is that you can provide different `config` values at deployment time, which will change the behaviour of writing the library during `_on_storage_attached`. Both cases (`library-write` and `storage-attached`) lead to the same method, which does evaluate the config, and it's also evaluated in `collect-status`.

I actually don't know whether I need to observe `config-changed` in order for `collect-status` to run, if not then this method could be removed entirely. It exists in the first place because it used to do more, and I thought it was worth keeping it around to document how configuration changes were handled.

tonyandrewmeyer replied:
collect-status will be triggered even if there are no other observers, so you don't have to have this. Keeping it because you plan to do with it seems fine (for a non-prod charm).

tonyandrewmeyer replied:
See also this issue: https://github.com/canonical/operator/issues/1265

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+            [
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
```
Could this be in the environment rather than an `export`?

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        format = cast(str, event.params[LIBRARY_INFO_FORMAT_PARAM])
+        match format:
+            case 'tree':
+                try:
+                    container.exec(['which', 'tree']).wait()
+                except ops.pebble.ExecError:
+                    logger.debug('_on_library_info: tree: installing dependencies')
+                    container.exec(['apt', 'update']).wait()
+                    container.exec(['apt', 'install', 'tree', '-y']).wait()
```
I find that a little weird.
You're ultimately in control of the docker image, so I would expect `tree` to be there from the start.

james-garner-canonical replied:
You're right. If I wrote a new `Dockerfile` for `calibre-web` in the future then I would include it. My current flimsy justification for `tree` is that it's only needed for this case of this action, so it's better to only install it if the user invokes it, but  I don't think that really holds up. Also there's another case where I do this for something that will be needed on every deployment ...

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
+                'cd /app/calibre-web',
```
Could this be `working-dir` in the layer?

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'ls-1': '\n'.join(stdout.lines)})
```
stdout seems to be bytes-like above, but is treated as str-like here.

james-garner-canonical replied:
Yeah I'm not sure what's going on there -- maybe it's actually a `typing.TextIO` if the process only outputs text?

dimaqq commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        )
+        command = f"bash -c '{c}'"
+        return {
+            'summary': f'{SERVICE_NAME} layer',
+            'description': f'pebble config layer for {SERVICE_NAME}',
```
Since these are text fields, it's odd to parametrise them. The content is arbitrary after all.

james-garner-canonical replied:
True, but I think if I changed the name of the service layer I would want these to match, and I'm not sure why there's anything wrong with parametresing here -- but I'd like to learn why not to do this if it doesn't feel right to you.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                    return
+                case 'clean':
+                    logger.debug(msg_prefix + 'cleaning ...')
+                    for fileinfo in contents:
```
Rather than iterating through each item in `/books/` removing them one by one, couldn't you just remove `/books/` (and then create an empty replacement) instead? It could be significantly fewer Pebble calls with a large library, I think.

james-garner-canonical replied:
This would be ideal, except that `/books/` is actually a mounted volume, so it can't be removed and recreated so easily.

tonyandrewmeyer replied:
Ah, right. If you continue with the charm, it might be worth putting the books folder into a folder in the mount so you can get this shortcut.

Might also be interesting reading this (closed, not complete) issue: https://github.com/canonical/operator/issues/1042

james-garner-canonical replied:
Thank you, helpful content in that issue. I think if I figure out how to configure the default library location in calibre-web automatically, bypassing the need for the user to write '/books/subfolder', then I'd just do the subfolder thing.

dimaqq reviewed (commented):
I've reviewed the first half only.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+                    for fileinfo in contents:
+                        container.remove_path(fileinfo.path, recursive=True)
+        logger.debug(f'_push_library_to_storage: {container=}')
+        library_path = Path(self.framework.model.resources.fetch('calibre-library'))
```
```suggestion
        library_path = Path(self.model.resources.fetch('calibre-library'))
```

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+            container.exec(['which', 'dtrx']).wait()
+        except ops.pebble.ExecError:
+            logger.debug('_push_and_extract_library: installing dependencies')
+            container.exec(['apt', 'update']).wait()
```
Same comment as earlier, that it seems like dependency installs would be better in pebble-ready.

james-garner-canonical replied:
Definitely agree, thank you for this advice.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+            ),
+        ]
+        container.exec(move_contents_up_one_level, working_dir=directory).wait()
+        container.exec(['rmdir', directory]).wait()  # error if not empty
```
Could this be `remove_path`?

james-garner-canonical replied:
It could be. During development I liked that `rmdir` would error if the directory wasn't empty, which helped me catch some issues that would have been difficult to debug otherwise, but now that everything is working, do you think it be better to switch to using `remove_path`?

tonyandrewmeyer replied:
Using recursive=False would get you those errors still, I believe. It seems the more 'native' way to do it, but exec is ok.

james-garner-canonical replied:
Oh yeah that's ideal, thanks

tonyandrewmeyer commented on tests/integration/test_charm.py:
```diff
@@ -0,0 +1,218 @@
+async def set_library_write_config(ops_test: OpsTest, behaviour: str) -> None:
+    logger.debug(f'set_library_write_config(ops_test, {behaviour=})')
+    model, app, unit = get_model_app_unit(ops_test)
+    model, app, unit = get_model_app_unit(ops_test)
```
Dupe line.

tonyandrewmeyer commented on tests/integration/test_charm.py:
```diff
@@ -0,0 +1,218 @@
+    await model.wait_for_idle(apps=[APP_NAME], status='active', raise_on_blocked=True, timeout=600)
+
+
+async def test_status_for_set_config(ops_test: OpsTest):
```
My inclination would be for this to be a parametrized test, but it's fine as-is too.

james-garner-canonical replied:
Good call, thanks

tonyandrewmeyer commented on tox.ini:
```diff
@@ -0,0 +1,86 @@
+    # if this charm owns a lib, uncomment "lib_path" variable
+    # and uncomment the following line
+    # codespell {[vars]lib_path}
+    codespell {tox_root}
```
This fails with Python < 3.11, although I have no idea why. Maybe it doesn't matter?

james-garner-canonical replied:
`codespell` came with `charmcraft init`, and only caused me pain by spellchecking files I didn't want it to (requiring config updates to ignore them), so it would be nice to have an excuse to remove it ...

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,257 @@
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
```
It only really matters if you're going to continue with this charm after onboarding, but `action-set` does have a size limit (I don't recall what it is offhand), and I think that would potentially get hit here with a large library. I assume this is more of an example action anyway, though?

james-garner-canonical replied:
Yeah I don't think this is a particularly useful action really, but it was fun to write. I'll look into how to handle `library-info` for larger libraries if I continue with this charm. Thanks.

tonyandrewmeyer commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+    type: filesystem
+    description: Storage mount for Calibre Library
+    location: /books
+    minimum-size: 1G
```
A lot of books, that's my kind of library :smile:.

tonyandrewmeyer commented on charmcraft.yaml:
```diff
@@ -0,0 +1,132 @@
+  charm:
+    plugin: charm
+    source: .
+  default-calibre-library:
```
I know that there have been changes with this in recent versions of charmcraft, but I thought that the charm plug-in would pull in all the files by default, except for whatever is in `.jujuignore`? I might be wrong or that might be specific to some charmcraft version, though.

james-garner-canonical replied:
By default, with no `parts`, I think it works as you say.

Once you add a part, in my case `default-calibre-library`, it will only pack that part, skipping *everything else* (`src/charm.py`, etc), which was a surprise for me. I can't find it now, but I seem to recall one piece of documentation (not [this one](https://juju.is/docs/sdk/charmcraft-yaml#heading--parts)) mentioning that after some juju (or charmcraft?) version, you'd have to provide the `charm` plugin explicitly. The documentation I looked at just now doesn't seem to mention this.

One thing I didn't figure out was why `charmcraft pack` now seems to pull in every file in my working directory, rather than just the known charm files and the ones specified in `parts`, which is how I thought it would work. I guess it all has to be set up in `.jujuignore`?

tonyandrewmeyer replied:
Yeah, I'm really unsure on what the situation is on this at the moment (and when it changed, etc). I don't think the docs do a great job of explaining this, although I know they are working on much more elaborate charmcraft docs, so hopefully that will solve it.

tonyandrewmeyer commented on README.md:
````diff
@@ -0,0 +1,60 @@
+juju run calibre-web/0 library-info format=tree --wait=1m
+```
+
+You can connect to the running `calibre-web` at `$APP_IP:8083`. The default credentials are `admin` and `admin123`. The library locations `/books` needs to be entered on the following screen before further options will be available.
````
> `admin123`

:scream_cat: 

A good action to add would be `set-credentials` or `add-user` or something like that, and to have the charm invalidate the awful `admin123` on `pebble-ready`.

Given that there are users, where are those stored? In some sort of DB in the container? If so then that probably ought to be in a storage mount instead, otherwise you'd lose all of this any time that K8s recreates the container, no?

james-garner-canonical replied:
An action like this was on the todo list! It does require figuring out how `calibre-web` stores its `config` and probably adding another storage mount like you suggest. I had weird issues with juju being very inconsistent with attaching the library storage, so I think that scared me away from adding a second storage volume till that was working consistently. Definitely not a production ready charm by any means!!

tonyandrewmeyer commented on README.md:
````diff
@@ -0,0 +1,60 @@
+juju run calibre-web/0 library-info format=tree --wait=1m
+```
+
+You can connect to the running `calibre-web` at `$APP_IP:8083`. The default credentials are `admin` and `admin123`. The library locations `/books` needs to be entered on the following screen before further options will be available.
````
When I did this, I ended up on a database configuration page that (if I understood correctly) wanted me to provide the path to the database. Is that something the charm could take care of? Or something I did wrong? I manually entered `/books/` and then was able to browse/read.

james-garner-canonical replied:
That is the expected flow currently. It's the flow that the docker image takes you through as well. I'd like the charm to automate it in future, if I continue working on it.

tonyandrewmeyer replied:
I somehow totally missed your "The library locations `/books` needs to be entered on the following screen before further options will be available", sorry.

But indeed, would be nice if the charm could automate this.

tonyandrewmeyer reviewed (commented):
Really nice! It's an impressive number of integration tests for an onboarding charm. I like that it's using (non OCI) resources, too - that's not that common, I think.

There's a lot of attention to type annotations here - I'm curious whether you did that because you felt it was most correct and/or whether it helped when you were actually writing the code & tests.

The integration tests are still running (`charmcraft test` is always very slow) but everything else I've tried works great!

tonyandrewmeyer commented:
My review collided with Dima's a bit, sorry :)

> The integration tests are still running (`charmcraft test` is always very slow) but everything else I've tried works great!

This ended up failing, as did a plain `tox -e integration`, oddly on building the charm even though doing that manually worked fine. Could just be my system?

james-garner-canonical commented:
@tonyandrewmeyer 
> > The integration tests are still running (`charmcraft test` is always very slow) but everything else I've tried works great!
> 
> This ended up failing, as did a plain `tox -e integration`, oddly on building the charm even though doing that manually worked fine. Could just be my system?

Just to confirm:

1. You can manually run `charmcraft pack` and that works?
2. You were able to deploy the charm and interact with it after that?
3. I didn't know about `charmcraft test`! But `tox -e integration` fails on `test_build_and_deploy` for you?

I've been running my integration tests with `tox -e integration -- --keep-models --model=calibre-web-model --no-deploy` for a while now, so it's possible that something is broken in `test_build_and_deploy`, and I'll look into that.

james-garner-canonical committed:
Remove tests/spread

This was added by charmcraft init, and I don't use it for testing. You
can run tox -e unit and tox -e integration

james-garner-canonical committed:
Remove skeletons from README

james-garner-canonical committed:
Document exact version of calibre-web docker image

In charmcraft.yaml and README.md

james-garner-canonical committed:
Remove placeholders in charmcraft.yaml

james-garner-canonical committed:
Cleanup charmcraft.yaml

james-garner-canonical committed:
Specify python version

james-garner-canonical committed:
Catch OSError on library-info if output is too large for juju result

james-garner-canonical committed:
Configure export and working-dir in pebble layer

Rather than with additional commands chained in the run command

james-garner-canonical committed:
Document why keeping _on_config_changed

james-garner-canonical committed:
Update copyright

james-garner-canonical committed:
Move exposing kubernetes port to install event

james-garner-canonical committed:
Use shorter lookups

james-garner-canonical committed:
Install dependencies in pebble_ready instead of on actions

james-garner-canonical committed:
Remove duplicated line

james-garner-canonical committed:
Use remove_path instead of exec rmdir

james-garner-canonical committed:
Follow logging conventions

james-garner-canonical committed:
Move initial _push_library_to_storage to _on_pebble_ready

Due to event ordering, the push was happening before dependencies were
installed in pebble_ready

james-garner-canonical committed:
Use a pythonic exception instead of go style error handling

james-garner-canonical committed:
Use process.wait_output instead of custom CaptureStdout class

james-garner-canonical committed:
Use exec find instead of iterating and calling remove_path

james-garner-canonical committed:
Parametrize test_config_changed

james-garner-canonical committed:
Parametrize test_status_for_set_config

james-garner-canonical committed:
Remove unit test that fails due to changes to _on_pebble_ready

james-garner-canonical committed:
Remove unused handlers

james-garner-canonical committed:
Partially parametrize test_library_actions

james-garner-canonical committed:
Remove spread.yaml

tonyandrewmeyer commented:
> 1. You can manually run `charmcraft pack` and that works?

Yes.

> 2. You were able to deploy the charm and interact with it after that?

Yes. My interaction was running the action to get the list in a couple of formats, and logging in and reading the one provided book.

> 3. I didn't know about `charmcraft test`!

It's the spread thing that is oddly hidden.

> But `tox -e integration` fails on `test_build_and_deploy` for you?

Yes. I just tried again and it still fails, but with:

```
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.056 :: E: Failed to fetch http://security.ubuntu.com/ubuntu/dists/jammy-security/InRelease  Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4000:1::101). - connect (101: Network is unreachable) Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4002:1::101). - connect (101: Network is unreachable) Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4000:1::102). - connect (101: Network is unreachable) Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4002:1::103). - connect (101: Network is unreachable) Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4000:1::103). - connect (101: Network is unreachable) Cannot initiate the connection to security.ubuntu.com:80 (2620:2d:4002:1::102). - connect (101: Network is unreachable) Could not connect to security.ubuntu.com:80 (185.125.190.82), connection timed out Could not connect to security.ubuntu.com:80 (185.125.190.81), connection timed out Could not connect to security.ubuntu.com:80 (91.189.91.83), connection timed out Could not connect to security.ubuntu.com:80 (91.189.91.82), connection timed out Could not connect to security.ubuntu.com:80 (91.189.91.81), connection timed out Could not connect to security.ubuntu.com:80 (185.125.190.83), connection timed out
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.056 :: E: Some index files failed to download. They have been ignored, or old ones used instead.
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.057 Failed to refresh package list: failed to run apt update.
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060 Traceback (most recent call last):
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060   File "/snap/charmcraft/4914/lib/python3.10/site-packages/craft_parts/packages/deb.py", line 449, in refresh_packages_list
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060     process_run(cmd)
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060   File "/snap/charmcraft/4914/lib/python3.10/site-packages/craft_parts/packages/deb.py", line 820, in process_run
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060     os_utils.process_run(command, logger.debug, **kwargs)
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060   File "/snap/charmcraft/4914/lib/python3.10/site-packages/craft_parts/utils/os_utils.py", line 345, in process_run
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060     raise subprocess.CalledProcessError(ret, command)
2024-09-19 18:36:03.448 :: 2024-09-19 18:36:03.060 subprocess.CalledProcessError: Command '['apt-get', 'update']' returned non-zero exit status 100.
```

I'm not sure if that was the error previously as well, but if it is then I assume that it's either a problem with my VM or with the upstream server. I'm unsure why `charmcraft pack` worked locally but maybe when pytest-operator does it it's using a different, slightly older image?
